packages/belos/tpetra/src/BelosTpetraOperator.hpp

```diff
@@ -0,0 +1,200 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+
+/*! \file BelosTpetraOperator.h
+    \brief This file provides a Tpetra::Operator interface so Belos can be integrated into
+     other codes as an abstract operator.
+*/
+
+#ifndef BELOS_TPETRA_OPERATOR_H
+#define BELOS_TPETRA_OPERATOR_H
+
+#include "Tpetra_MultiVector_decl.hpp"
+#include "Tpetra_Operator.hpp"
+#include "Tpetra_Map.hpp"
+
+#include "BelosLinearProblem.hpp"
+#include "BelosStatusTest.hpp"
+#include "BelosOutputManager.hpp"
+#include "BelosTpetraAdapter.hpp"
+#include "BelosSolverFactory.hpp"
+
+#include "Teuchos_ParameterList.hpp"
+
+/*! \class Belos::TpetraOperator
+    \brief This class provides an interface to the Tpetra::Operator class, so Belos can be 
+    integrated into other codes as an abstract operator.
+*/
+
+namespace Belos {
+
+///////////////////////////////////////////////////////////////
+//-------- class BelosTpetraOperator --------------------
+//
+// This class will allow a Belos solver to be called as an Tpetra_Operator.
+// Thus, it can use itself as a preconditioner if need be.  It can
+// also be used as the inner iteration of Anasazi :)
+//
+///////////////////////////////////////////////////////////////
+template<class Scalar = ::Tpetra::Details::DefaultTypes::scalar_type,
+         class LocalOrdinal = ::Tpetra::Details::DefaultTypes::local_ordinal_type,
+         class GlobalOrdinal = ::Tpetra::Details::DefaultTypes::global_ordinal_type,
+         class Node = ::Tpetra::Details::DefaultTypes::node_type>
+class TpetraOperator : public virtual Tpetra::Operator<Scalar, LocalOrdinal, GlobalOrdinal, Node> {
+public:
+  //! @name Typedefs
+  //@{
+  //! The Tpetra Multivector 
+  typedef Tpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node> Tpetra_MultiVector;
+  //! The Tpetra Operator
+  typedef Tpetra::Operator<Scalar, LocalOrdinal, GlobalOrdinal, Node> Tpetra_Operator;
+  //! The Tpetra Map
+  typedef Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> Tpetra_Map;
+
+  //! @name Constructor / Destructor
+  //@{ 
+  
+  //! Constructor
+  TpetraOperator( const Teuchos::RCP<LinearProblem<Scalar,Tpetra_MultiVector,Tpetra_Operator> >& lp, 
+		  const Teuchos::RCP<Teuchos::ParameterList>& plist,
+                  bool initSolnVec = false )  
+  : lp_(lp), 
+    plist_(plist),
+    initSolnVec_(initSolnVec)
+  {
+  // Get the solver's name from the parameter list, use block Gmres by default.
+  std::string solver = plist_->get("Solver", "BlockGmres");
```

Comment:
If you're going to pass the ParameterList to the solver, then it's best to have the solver name be stored separately from that ParameterList.  Otherwise, the solver's setParameters method might report an error, because it doesn't understand the "Solver" parameter.

Reply:
So then this would be the reason to delete the solver name from the parameter list....